### ISSUE TITLE
support the new emoji panel in windows 10

### ISSUE THREAD

**Issue by mohdshara** [nvaccess/nvda#7273]
### Steps to reproduce:
In windows 10 windows insider build 16215 Microsoft introduced the Emoji panel. 
1. focus on any text field and then press the windows key + ; or + . 
2. the emoji panel appears. 
### Expected behavior:
NVDA doesn't speak anything, and moving with the right and left keys moves between Emojies but NVDA doesn't read any of them. if you press enter the emoji would get entered as expected and announced.
### Actual behavior:
NVDA doesn't tell us that the panel has opned nor does it read when moving between items. Note that Narrator reads those fine as expected.
### System configuration:
NVDA version:
NVDA Version: next-14106,059e7f02

NVDA Installed or portable:
installed

Other information:
none needed

Windows version:
Windows Version 1703 (OS Build 16215.1000)

the other questions in the template don't apply here.

**Comment by josephsl** [comment]
Hi, partly because we need to find out if the new XAML touch input panel fires useful events and announce changes if any. CC @MichaelDCurran

**Comment by jcsteh** [comment]
I looked into this briefly a couple of days ago. Note that this can be used even if you don't use the touch keyboard.

The emoji panel does work with Narrator.

The emoji panel is beneath a top level window named "Microsoft Text Input Application". [This MSDN article about IMEs in Windows Store apps](https://msdn.microsoft.com/en-us/library/windows/apps/hh967425.aspx#accessibility) says the following about accessibility:

> Implement the following convention to make your IMEs conform to the accessibility requirements and to work with Narrator. To make candidate lists accessible, your IMEs must follow this convention.
> 
> - The candidate list must have a UIA_AutomationIdPropertyId equal to "IME_Candidate_Window" for lists of conversion candidates or "IME_Prediction_Window" for lists of prediction candidates.
> - When the candidate list appears and disappears, it raises events of type UIA_MenuOpenedEventId and UIA_MenuClosedEventId, respectively
> - When the current selected candidate changes, the candidate list raises a UIA_SelectionItem_ElementSelectedEventId. The selected element should have a property UIA_SelectionItemIsSelectedPropertyId equal to TRUE.
> - The UIA_NamePropertyId for each item in the candidate list must be the name of the candidate. Optionally, you can provide additional information to disambiguate candidates through UIA_HelpTextPropertyId.

However, it seems Microsoft doesn't follow their own guidelines here with regard to automation ids:

1. There are two lists in that window. The first list has an automation id of TEMPLATE_PART_ExpressionFullViewItemsGrid. The second has an automation id of TEMPLATE_PART_ExpressionFullViewGroupsList.
2. The first list is the list of emoji. The name of each item in this lit is the Unicode representation of the emoji.
3. The selected item in the first list does change when you use the arrow keys. That means it probably does raise selection events. I think Narrator probably listens for selection events globally, which explains how that works.
4. The second is the list of categories.Unfortunately, the category list is all private use (and thus not standardised) Unicode characters.
5. When you press tab, the selected item in the second list changes (corresponding to the selected category).
6. Note that Narrator doesn't report the categories in any way.

CC @MichaelDCurran.

**Comment by josephsl** [comment]
Hi, yes, element selected event is indeed raised. I’ll come up with a temporary workaround in the WinTenApps add-on. Thanks.

**Comment by josephsl** [comment]
Hi,

Technical: these are "floating suggestion items" in that they are not part of the app in question. Rather, they appear as part of a floating window with its own app executable name. Numerous solutions come to my mind:

* #6241: I can add a new FloatingSuggestionsListItem class that will do what SuggestionsListItem does except for checking for controller for event (this is the solution employed in latest WinTenApps snapshot as of June 12th).
* A request event for UIA objects should be added for the app module for the new XAML input panel. Among other things, this opens up numerous possibilities such as working with floating windows.

Important note: the implementation @JCSTeh described and I wrote a workaround for can change in future Insider builds.

Thanks.

**Comment by josephsl** [comment]
Hi,

Although the suggestions list item is in master, I'll hold off until at least two slow ring builds of Fall Creators Update Insider Preview are released in order to gauge how it'll work in the wild. Thanks.

**Comment by mohdshara** [comment]
in build 16237 some changes were made to the Emoji panel, in particular this may be beneficial to the implementation of this ticket:

"• Have you ever wondered “What the heck is this emoji supposed to be”? Wonder no more! If you hover your mouse over any of the emoji in the Emoji Panel, you’ll now see a handy showing you the exact Unicode character name. Bonus: we’ve added tooltips to the categories so they’re no longer a mystery."

**Comment by PratikP1** [comment]
Experientially speaking, in build 16237, there appears to be no difference in how Narrator or NVDA treat the emoji panel than originally reported by @jcsteh. I filed a bug in the Feedback Center last week. See the following. https://aka.ms/fk2cna

**Comment by josephsl** [comment]
Hi, and a possible modification might be in order, as the panel title is announced when I use arrow keys (at least the implementation in my add-on may require some edits). Thanks.

**Comment by PratikP1** [comment]
Here is a log snippet of what NVDA says when Win 10 Apps Essential 2017-07-05-dev is running. Window Title is not announced There’s nothing without the add-on running.

IO - inputCore.InputManager.executeGesture (12:20:07.668):
Input: kb(laptop):windows+.
IO - inputCore.InputManager.executeGesture (12:20:09.502):
Input: kb(laptop):tab
IO - speech.speak (12:20:09.595):
Speaking [LangChangeCommand (u'en_US'), u'1 inches']
IO - speech.speak (12:20:09.693):
Speaking [LangChangeCommand (u'en_US'), u'blank']
IO - speech.speak (12:20:10.374):
Speaking [LangChangeCommand (u'en_US'), u'\U0001f600  1 of 256']
IO - inputCore.InputManager.executeGesture (12:20:12.174):
Input: kb(laptop):shift+tab
IO - speech.speak (12:20:12.278):
Speaking [LangChangeCommand (u'en_US'), u'1 inches']
IO - speech.speak (12:20:12.394):
Speaking [LangChangeCommand (u'en_US'), u'blank']
IO - speech.speak (12:20:12.746):
Speaking [LangChangeCommand (u'en_US'), u'\U0001f602  1 of 45']
IO - inputCore.InputManager.executeGesture (12:20:14.892):
Input: kb(laptop):rightArrow
IO - speech._speakSpellingGen (12:20:15.151):
Speaking character u'line feed'
IO - speech.speak (12:20:15.158):
Speaking [LangChangeCommand (u'en_US'), u'\U0001f600  2 of 45']
IO - inputCore.InputManager.executeGesture (12:20:15.964):
Input: kb(laptop):rightArrow
IO - speech._speakSpellingGen (12:20:16.204):
Speaking character u'line feed'
IO - speech.speak (12:20:16.213):
Speaking [LangChangeCommand (u'en_US'), u'\U0001f601  3 of 45']
IO - inputCore.InputManager.executeGesture (12:20:17.252):
Input: kb(laptop):rightArrow
IO - speech._speakSpellingGen (12:20:17.484):
Speaking character u'line feed'
IO - speech.speak (12:20:17.492):
Speaking [LangChangeCommand (u'en_US'), u'\U0001f923  4 of 45']
IO - inputCore.InputManager.executeGesture (12:20:17.948):
Input: kb(laptop):tab
IO - speech.speak (12:20:18.045):
Speaking [LangChangeCommand (u'en_US'), u'1 inches']
IO - speech.speak (12:20:18.147):
Speaking [LangChangeCommand (u'en_US'), u'blank']
IO - speech.speak (12:20:18.783):
Speaking [LangChangeCommand (u'en_US'), u'\U0001f600  1 of 256']
IO - inputCore.InputManager.executeGesture (12:20:20.069):
Input: kb(laptop):tab
IO - speech.speak (12:20:20.177):
Speaking [LangChangeCommand (u'en_US'), u'1 inches']
IO - speech.speak (12:20:20.285):
Speaking [LangChangeCommand (u'en_US'), u'blank']
IO - speech.speak (12:20:21.020):
Speaking [LangChangeCommand (u'en_US'), u'\U0001f469  1 of 175']
IO - inputCore.InputManager.executeGesture (12:20:22.428):
Input: kb(laptop):tab
IO - speech.speak (12:20:22.526):
Speaking [LangChangeCommand (u'en_US'), u'1 inches']
IO - speech.speak (12:20:22.638):
Speaking [LangChangeCommand (u'en_US'), u'blank']
IO - speech.speak (12:20:23.562):
Speaking [LangChangeCommand (u'en_US'), u'\U0001f388  1 of 273']
IO - inputCore.InputManager.executeGesture (12:20:25.029):
Input: kb(laptop):end
IO - speech._speakSpellingGen (12:20:25.796):
Speaking character u'line feed'
IO - inputCore.InputManager.executeGesture (12:20:27.059):
Input: kb(laptop):leftArrow
IO - speech._speakSpellingGen (12:20:27.286):
Speaking character u'line feed'
IO - speech.speak (12:20:27.296):
Speaking [LangChangeCommand (u'en_US'), u'\u23f2  272 of 273']

**Comment by jcsteh** [comment]
@josephsl, I think this would be good to support in core. If you'd be willing to submit a PR, that'd be great. We should think about the best implementation, though:

[**@josephsl**](https://github.com/josephsl) commented on [12 Jun 2017, 12:05 GMT+10](https://github.com/nvaccess/nvda/issues/7273#issuecomment-307674396 "2017-06-12T02:05:14Z - Replied by Github Reply Comments"):
> * #6241: I can add a new FloatingSuggestionsListItem class that will do what SuggestionsListItem does except for checking for controller for event (this is the solution employed in latest WinTenApps snapshot as of June 12th).

1. Any thoughts on the likelihood of other floating suggestions like this?
2. How do you determine what should get this behaviour? Do you use the automation id or similar or is it much more generic? I'd be concerned about unintentional reporting if it were too generic.
3. Is "suggestions" the right way to think about this? This is probably more like candidates in an input method.

> * A request event for UIA objects should be added for the app module for the new XAML input panel.

Does this need requestEvent? Is it blocked by shouldAcceptEvent by default?

I'm inclined to go more for an app module specific implementation here. Aside from the open questions above, handling the categories properly will require specific code, I think.

**Comment by josephsl** [comment]
Hi,

App module method: I'll report back if I find anything new (using slow ring 16232 on my physical desktop, which should make it easier for me to test this).
Candidate or not: I see why it could be labeled that way. If so, I guess it makes implementation super easy unless @MichaelDCurran says otherwise.
Other floating suggestions: not that I know of. I'll ask a contact from MS Shell team about this, as well as asking if the Emoji panel implementation is stable, as the answer to that changes when the PR would be ready; if the panel code is stable, I can clean things up and send a PR as early as Tuesday, if not, I can wait until September.

Thanks.

**Comment by PratikP1** [comment]
I wonder if there is progress in adding emoji panel support to NVDA. @josephsl's Win10Apps add-on provides an excellent start. I believe that Microsoft uses a similar mechanism to support for word suggestions for hardware keyboards, a feature introduced in insider build 17035.

**Comment by josephsl** [comment]
Hi, as much as I want to see this integrated, it is risky due to ever-changing nature of Windows 10. Thanks.

**Comment by PratikP1** [comment]
This is a question for the general community. When do we consider new Windows features stable? It appears that Microsoft is continuously tweaking the OS now. Not having access to important features such as the emoji input panel or the newly introduced timeline because of the possibility that Microsoft will be changing it continuously could make us go down the road of not supporting important features. Is it that we need to wait till the feature is included without change in several major releases of Windows? With that logic, NVDA may be falling behind significantly.

**Comment by josephsl** [comment]
Hi, part of this has to do with our current review procedure in that we need to wait at least two weeks for things to make it from PR to master once accepted. During that time, things can change, thus requiring edits. This is precisely the reason why the add-on mentioned exists: to make sure NVDA supports latest features on time. At the moment there’s no work to be done for Timeline, as NVDA already supports it. Thanks.

**Comment by PratikP1** [comment]
What about the emoji panel? It seems to have been fairly stable for quite a few builds. I was using Timeline as an illustration and not specifically. Thanks for all your work.

**Comment by josephsl** [comment]
Hi,

Emoji panel seems stable, but things can change. I'm about to publish a PR for it, hoping it can be included in 2018.1.

Thanks.

**Comment by josephsl** [comment]
Hi,

Actually, there is a PR: #7378. Not reviewed in months, partially due to changes during that time. Now that it is stable, I'd like to request review again. Thanks.

**Comment by PratikP1** [comment]
Again, Thank you. The only possible issue with Win10Apps implementation seems to be that NVDA doesn’t announce that the panel has popped up.

**Comment by josephsl** [comment]
Hi, it uses menu open event, which is a bit questionable at times. If it used window open event, then I can fix it tonight. Thanks.

**Comment by PratikP1** [comment]
I can request window open event for the emoji panel. I can also ask them to make that the case for the hardware keyboard suggestions panel if that’s what they are using.

**Comment by josephsl** [comment]
Hi, actually, window open event is fired. I need to make sure this is the case for 1709, and if so, it’ll be a matter of supporting that event and item selected event. Thanks.

**Comment by josephsl** [comment]
Hi,

Testing shows the following:

1. Window open event is fired.
2. Item selected event is used.
3. Window opened event isn't fired in 16299 (Version 1709).

In this case, I'll come up with a snapshot of my add-on that'll resolve item announcement bug for 17000 series. Thanks.

**Comment by armick** [comment]
Why left/right/back space/delete does not announce anything? Both JAWS/Narrator does. Should you F*K something when adding/improving something? I'm crying for my $120 donations for the past 4 releases!

**Comment by josephsl** [comment]
Hi,

Okay, let's grab something to drink (perhaps soda or wine) and have a deep talk about what's happening. I'm listening - I won't say anything until you tell me the whole thing. Although I think NV Access people might not take it easily, I'm willing to listen to whatever that's in your mind, including rants, insults, and curses.

So, my friend, please tell me what's really bothering you about NVDA right now. Is it about bugs? Something wrong with emoji panel support or something else? Or, if you don't want to talk about what's really going on in public, that's fine.

Sincerely, from a really concerned NVDA contributor

**Comment by josephsl** [comment]
Hi,

Hmmm, @armick, is that the issue you're seeing, or something else? If it is exactly what Quentin is saying, chances are that someone will report this to Microsoft.

Technical: I think Quentin is refering to build 18298 (fast ring). If so, this might be caused by UIA window open event not being fired and/or detected by NVDA (I tested this scenario numerous times to no avail). Usually, re-initializing UIA handler and killing modern keyboard app (WindowsInternal.ComposableShell.Experiences.TextInput.InputApp.exe) via Task Manager should resolve this, but apparently, that workaround isn't working for me in build 18298. This also affects cloud clipboard paste command; strangely, dictation and hardware keyboard input suggestions are working. I consider emoji panel support failure to be a high severity bug for Windows 10 users.

However, if your issue is something other than emoji panel support, then please do let us know what's going on and how to reproduce the problem you're seeing.

Thanks.

**Comment by Qchristensen** [comment]
Sorry, I should have clarified, I did mean insider build of Windows, but I am using Slow insider build 18272 (which expires in a couple of days apparently.  Ridiculous that you get builds expiring, or even messages about builds expiring when there isn't a new build on that ring available.  But I digress :)  The issue I'm seeing with the emoji panel definitely seems to occur whether or not NVDA is running, so that seems like it might be a different issue to Joseph's.
